client/src/components/BlockEditor/BlockEditorEditing.tsx

```diff
@@ -176,7 +186,31 @@ export class BlockEditorEditingPresentational extends React.Component<
     }
   };
 
-  private onKeyDown = event => {
+  private onKeyDown = (event: any, change: Change) => {
+
+    const isMovingLeft = event.key === "ArrowLeft" || event.key === "Backspace";
+    const isMovingRight = event.key === "ArrowRight";
+
+    // simulate the inteded move to the left or right
+    // because they are simulated we don't use the original change object
+
+    let valueAfterSimulatedChange = change.value;
+
+    if (isMovingLeft) {
+      valueAfterSimulatedChange = valueAfterSimulatedChange.change().move(-1).value;
+    }
+
+    if (isMovingRight) {
+      valueAfterSimulatedChange = valueAfterSimulatedChange.change().move(1).value;
+    }
+
+    if (isCursorInPotentiallyProblematicPosition(valueAfterSimulatedChange)) {
+      const correctedChange = handleMovingCursor(change, valueAfterSimulatedChange, isMovingLeft, isMovingRight);
+      this.props.updateBlock({ id: this.props.block.id, value: correctedChange.value, pointerChanged: false });
+      event.preventDefault();
+      return false;
+    }
```

Comment:
In the very least I think the above should mostly be in a separate function or similar. I kind of get what it's doing, but think it could be further cleaned up.

Reply:
This also seems like stuff I could imagine being put into a Slate library or similar. Most of the other contents of this file have to do with things really orthogonal to the editing process itself; like general component mounting, updating the database, and updating the redux stores. Much of the above seems much more text-editor specific.